#160577438 reset password

ibetween opened the pull request:
## What does this PR do?
- The user should be able to reset password

## Description of Task to be completed?
- Users will input their email
- The user will be provided with a link to reset their password
## How should this be manually tested?
- Clone this branch ft-signup-160577437
- Run the following commands on your terminal:
  `yarn install` - to install required dependencies
  `yarn start` - to start server
- When the server starts, you will be redirected to the browser
- Navigate to this url: [http://localhost:3000/forgot_password](http://localhost:3000/forgot_password)
<img width="1440" alt="screen shot 2018-11-08 at 12 27 38" src="https://user-images.githubusercontent.com/29597869/48189678-cadae780-e351-11e8-8403-a42716175e92.png">
- An email will be sent.
- Click the email and reset your password
<img width="1422" alt="screen shot 2018-11-08 at 12 33 44" src="https://user-images.githubusercontent.com/29597869/48190002-93b90600-e352-11e8-823b-1fa9a95e7b9d.png">


## Any background context you want to provide?
- The link will expire the password has been reset
## What are the relevant pivotal tracker stories?
- https://www.pivotaltracker.com/story/show/160577438
## Packages installed
- validator: will validate the email
## Deployment note
N/A

## Related PRs branch | PR (branch_name) | (pr_link)
N/A

## Todos
- [x] Raise PR
- [ ] Test
- [ ] Documentation


## Screenshots (if appropriate)
- Screenshots have been provided above
## Questions:
<to be deleted>

[sample pr](https://github.com/andela/ah-shakas/pull/39)

Commit:
feat(reset-password): - Reset password

- Users can receive an email with a link to reset their password.
- Implement Error validation
- Users can successfully reset their password.

Commit:
feat(reset-password): - Reset password

- Users can receive an email with a link to reset their password.
- Implement Error validation
- Users can successfully reset their password.

Commit:
feat(reset-password): - Reset password

- Add reducer tests

[WIP #160577477]

Commit:
feat(reset-password): - Fix Es6 rules

- Fix single quotes for the strings warning
- Fix trailing comma-dangle warning

[WIP #160577438]

Commit:
feat(reset-password): - Fix Es6 rules

- Fix single quotes for the strings warning
- Fix trailing comma-dangle warning

[WIP #160577438]

Commit:
feat(reset-password): - Refactor style

- Fix Eslint errors
- Refactor style

[WIP #160577438]

Commit:
feat(reset-password): - Refactor style

- Fix Lint errors errors

[WIP #160577438]

Arthur236 commented on src/__tests__/reducers/forgot.password.reducer.test.js:
```diff
@@ -0,0 +1,29 @@
+import restRequest from '../../reducers/reset.password.reducer';
+import RESPONSES from '../../mock/responses';
+
+const initialState = {
```
Initial state should be in its own file i.e. `initialState.js` which will have different state objects.

Arthur236 commented on src/actions/password.reset.action.js:
```diff
@@ -0,0 +1,17 @@
+import axios from 'axios';
+import { ERROR_ACTION, RESET_PASSWORD_ACTION } from '../constants';
+
+const resetPasswordRequest = ({ email }) => dispatch => axios.post('http://127.0.0.1:8000/api/users/email_sent', { email })
```
Have this part of the url as an environment variable like so:
`REACT_APP_BASE_URL=http://127.0.0.1:8000/api/`

Arthur236 commented on src/components/Auth/ForgotPasswordComponent.js:
```diff
@@ -0,0 +1,29 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import { connect } from 'react-redux';
+import { bindActionCreators } from 'redux';
```
Let's not use bindActionCreators and mapDispatchToProps. Just pass in an object containing all your actions as the second parameter in the `connect` function

ibetween replied:
Alright

Arthur236 commented on src/components/Auth/ForgotPasswordComponent.js:
```diff
@@ -0,0 +1,29 @@
+  resetPasswordRequest: PropTypes.func.isRequired,
+};
+
+const mapStateToProps = ({ reset }) => ({ ...reset });
```
Move this part into a container component as discussed earlier today

Commit:
feat(reset-password): - Test reducers and actions

- Add the tests for the resetting password reducer
- Add tests for the reset password and resetting password actions
[WIP #160577438]

Commit:
feat(reset-password): - Test actions and components

- Add the tests for the forgot password component
- Add tests for the resetting password action
[WIP #160577438]

Arthur236 commented:
@james-chege Resolve your conflicts

Commit:
feat(reset-password): - Refactor code

- Obey the set conventions
[WIP #160577438]

Commit:
feat(reset-password): - Rebase

- Rebase against develop
[WIP #160577438]

Commit:
feat(reset-password): - Fix Linting warnings

- fixed linting errors
[FINISHES #160577438]

Commit:
feat(reset-password): - Redirect after success full reset

- user will be redirected after successful password reset
[FINISHES #160577438]

ibetween commented:
> @james-chege Resolve your conflicts

Hi @Arthur236 I have resolved the conflicts

Arthur236 commented:
Kindly work on the hound violations @james-chege

Arthur236 commented:
Delete this file `package.json`

Arthur236 commented:
Kindly follow the naming conventions that we agreed upon
Almost all your files are incorrectly named

Use camel case for files like this `reseting.password.action.js`. Hence the correct naming should be `resetingPassword.action.js` and the same will apply for the test `resetingPassword.action.test.js`

Arthur236 commented:
Also I see you have created a `__tests__` directory. Why did you decide to do this when we already had a `__test__` directory

ibetween commented:
Thanks @Arthur236 I had discussed with team and regarding to the documentation it states it should be `tests` so this is why I changed it. Do I change it back to `test`?

ibetween commented:
> Delete this file `package.json`

I deleted the `package-lock.json` this is the one with the packages

Commit:
feat(reset-password): - Fix conventions

- Fix the recommended conventions
[FINISHES #160577438]

Commit:
feat(reset-password): - Fix Hound issues

- Fix the the hound issues
[FINISHES #160577438]

Commit:
feat(reset-password): - Fix Hound issues

- Fix the the hound issues
[FINISHES #160577438]

Arthur236 commented:
> Thanks @Arthur236 I had discussed with team and regarding to the documentation it states it should be `tests` so this is why I changed it. Do I change it back to `test`?

No. Leave it as tests but move all the other tests that were in the old folder to that new one and only have tests

ibetween commented:
> > Thanks @Arthur236 I had discussed with team and regarding to the documentation it states it should be `tests` so this is why I changed it. Do I change it back to `test`?
> 
> No. Leave it as tests but move all the other tests that were in the old folder to that new one and only have tests

Alright. I have moved them.

Arthur236 commented:
Your reducers are still not renamed correctly

Arthur236 commented:
<img width="1509" alt="screenshot 2018-11-13 at 15 33 00" src="https://user-images.githubusercontent.com/31339738/48413660-7375b680-e759-11e8-9ca2-c57f41d8b483.png">

Why do you have 2 navbars in your homepage?

Arthur236 commented:
<img width="1513" alt="screenshot 2018-11-13 at 15 36 19" src="https://user-images.githubusercontent.com/31339738/48413805-e7b05a00-e759-11e8-95e8-4aa804a69005.png">

Have a title on this page so that its more obvious what the user is supposed to do.
Also make the form a bit smaller

Arthur236 commented:
<img width="1511" alt="screenshot 2018-11-13 at 15 38 26" src="https://user-images.githubusercontent.com/31339738/48413900-30681300-e75a-11e8-8bef-16a0700c129d.png">

Apply the above comments on this page as well

In addition to that, change the placeholder text from `confirm` to `Confirm Password`

ibetween commented:
> Your reducers are still not renamed correctly

alright. Leme check it.

ibetween commented:
> <img alt="screenshot 2018-11-13 at 15 38 26" width="1511" src="https://user-images.githubusercontent.com/31339738/48413900-30681300-e75a-11e8-8bef-16a0700c129d.png">
> 
> Apply the above comments on this page as well

Alright.

ibetween commented:
> <img alt="screenshot 2018-11-13 at 15 33 00" width="1509" src="https://user-images.githubusercontent.com/31339738/48413660-7375b680-e759-11e8-9ca2-c57f41d8b483.png">
> 
> Why do you have 2 navbars in your homepage?

Haden't noticed this leme rectify it.

Commit:
feat(reset-password): - Fix conventions

- Fix naming conventions for reducers
[FINISHES #160577438]

Commit:
feat(reset-password): - Fix Style

- Fix broken style
[FINISHES #160577438]

Commit:
feat(reset-password): - Fix Style

- Modify the pages inline errors
[FINISHES #160577438]

Commit:
feat(reset-password): - Fix Style

- fix a typo
[FINISHES #160577438]

Commit:
feat(reset-password): - Fix Hound errors

- fix scss errrors
[FINISHES #160577438]

Commit:
feat(reset-password): - Fix Hound errors

- fix scss errrors
[FINISHES #160577438]

Commit:
feat(reset-password): - Fix Hound errors

- Fix Properties order
[FINISHES #160577438]

Commit:
feat(reset-password): - Use environment variables

- Put the base link in the environment
[FINISHES #160577438]

Commit:
feat(reset-password): - Toast a reset messages

- Toast a message on password reset success.
[FINISHES #160577438]

Commit:
feat(reset-password): - Fix tests

- Fix the failing action testss
[FINISHES #160577438]

Commit:
feat(reset-password): - Fix tests

- Fix the failing action tests
[FINISHES #160577438]

Arthur236 commented on src/reducers/resettingPassword.reducer.js:
```diff
@@ -0,0 +1,23 @@
+
+const { RESETTING_PASSWORD_ACTION, RESETTING_ERROR_ACTION } = CONSTANTS;
+
+const initialState = {
```
Use the initial state in the `initialState.js` file.
Don't define it here

Arthur236 commented on src/reducers/resetPassword.reducer.js:
```diff
@@ -0,0 +1,23 @@
+
+const { RESET_PASSWORD_ACTION, ERROR_ACTION } = CONSTANTS;
+
+const initialState = {
```
Use the initial state in the initialState.js file.
Don't define it here

Commit:
feat(reset-password): - Fix eslint rule

- Fix eslint line break rules
[FINISHES #160577438]

Commit:
feat(reset-password): - Modify initial state

- Put initial state in the recommended file.
[FINISHES #160577438]

Arthur236 commented on src/actions/passwordReset.action.js:
```diff
@@ -0,0 +1,19 @@
+
+const { ERROR_ACTION, RESET_PASSWORD_ACTION } = CONSTANTS;
+
+const resetPasswordRequest = ({ email }) => dispatch => axios.post(`${process.env.REACT_APP_BASE_URL}/users/email_sent`, { email })
```
Hadn't noticed this earlier
Use the axios instance: `api` that was created in a previous PR to make your api calls

Arthur236 commented on src/actions/resetingPassword.action.js:
```diff
@@ -0,0 +1,20 @@
+
+const { RESETTING_ERROR_ACTION, RESETTING_PASSWORD_ACTION } = CONSTANTS;
+
+const ResettingPassword = (token, password) => dispatch => axios.put(`${process.env.REACT_APP_BASE_URL}/users/password_reset?token=${token}`, { password })
```
Same case here. Use `api`

Arthur236 commented on src/__tests__/actions/resettingPassword.action.test.js:
```diff
@@ -0,0 +1,47 @@
+  const mock = new MockAdapter(axios);
+  const token = 'xaAaddddAD3323@#@$$@%@';
+  const password = 'AverysecurePassword';
+  const url = `${process.env.REACT_APP_BASE_URL}/users/password_reset?token=${token}`;
```
You don't need to attach the base url here. Just have it as 
```
const url = `users/password_reset?token=${token}`;
```

Commit:
feat(reset-password): - Fix hound

- Ignore function-paren-newline" rule
[FINISHES #160577438]

Commit:
feat(reset-password): - Fix structure

- Use api instead of axios
[FINISHES #160577438]

Commit:
feat(reset-password): - Fix structure

- Use api instead of axios in resetting password
[FINISHES #160577438]

Arthur236 commented:
<img width="492" alt="screenshot 2018-11-14 at 16 27 44" src="https://user-images.githubusercontent.com/31339738/48485554-3d543780-e82a-11e8-94e2-8a6e43c14cf1.png">

Change the placeholder text from `confirm` to `Confirm Password`

Commit:
feat(reset-password): - Fix style

- Fix password resetting form input placeholder
[FINISHES #160577438]

Arthur236 commented on src/components/Auth/ForgotPasswordComponent.jsx:
```diff
@@ -0,0 +1,25 @@
+import ForgotPasswordForm from '../forms/ForgotPasswordForm';
+
+class ForgotPasswordComponent extends React.Component {
+  state ={
```
Since you're not using state anywhere, you can remove this declaration

Arthur236 commented on src/components/Auth/ResetPasswordComponent.jsx:
```diff
@@ -0,0 +1,33 @@
+import ResetPasswordForm from '../forms/ResetPasswordForm';
+
+class ResetPasswordComponent extends React.Component {
+  state ={
```
Since you're not using state anywhere, you can remove this declaration
